### PR TITLE
Inference flavor experiments

### LINKED ISSUE #4: Create separate test requirements in `pyproject.toml`
Currently, all `dev` dependencies are installed in the test workflow. However, these come with packages like `sphinx` which are not required for the actual workflow. We should create a separate test dependency list to avoid the overhead from installing unnecessary packages during the test action.

### PR DESCRIPTION
**Pull request by mcw92**
- This PR introduces job script generation scripts for the inference flavor experiments. 
- Due to OOM errors, I refactored the code to decouple the training completely from potentially building a globally shared model and evaluation. As training the classifiers is the computationally most expensive step, the locally trained subforests are dumped to file directly after the training. This prevents loss of an alredy trained model due to subsequent errors during building the globally shared model or evaluation.
- I added new functionality, including an exemplary script and test, to evaluate in parallel from pickled model checkpoints.
- I updated the test action to use the Codecov app instead of generating and committing the badge locally using a bot. In addition, I introduced matrix testing for supported Python versions and OS types.
- I introduced separate test dependencies.
- I added analysis scripts for plotting inference comparison results.
- I added scripts to generate job scripts for each experiment series. Even though this introduces some redundancy, this makes handling the experiments easier for now.

Fixes #4.

### CODE REVIEW

File: scripts/analysis/find_node_overlap.py
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/find_node_overlap.py:1]: This was just a useful helper script to identify broken nodes as the overlap of the lists of nodes used by jobs that failed. I further refined the overlap by comparing it to the node lists of jobs that ran successfully. In this way, I could remove functional nodes from the overlap. I will add a docstring and explaining comments to the script.

File: scripts/analysis/acc_drop.py [hunk 1]
```diff
@@ -52,19 +53,46 @@
                     global_test_accuracy
                 )
 
+print(results)
+
+for dirpath, dirnames, filenames in os.walk(root_dir):
+    for filename in filenames:
+        if filename.endswith(".out"):
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/acc_drop.py]: I worked your suggestion into the code. However, I did not apply your suggested change directly as this change required some furhter modifications in the code.

File: scripts/examples/evaluate_from_checkpoint_parallel_synthetic.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    if comm.rank == 0:
+        log.info(
+            "**********************************************************************\n"
+            "* Multi-Node Random Forest Classification of Balanced Synthetic Data *\n"
+            "**********************************************************************\n"
+            f"Hyperparameters used are:\n{args}"
+        )
+
+    # Train distributed random forest on balanced synthetic classification data.
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/examples/evaluate_from_checkpoint_parallel_synthetic.py]: I corrected the comment in the code.

File: tests/test_rf_parallel_synthetic.py [hunk 3]
```diff
@@ -87,6 +90,6 @@ def test_parallel_synthetic(random_state_model: int) -> None:
         experiment_id=experiment_id,
         save_model=save_model,
     )
-    if comm.rank == 0:
-        log_file.unlink()
-        shutil.rmtree(output_dir)
+    comm.barrier()
+    # Remove all files generated during test in temporary directory.
+    shutil.rmtree(str(mpi_tmp_path), ignore_errors=True)
```
[the remark this answers is not in the document]
Comment (mcw92) [reply tests/test_rf_parallel_synthetic.py]: I introduced a fixture to clean up MPI-enabled temporary paths in Pytest. The fixture is defined in confest.py and used in all parallel tests. The normal `tmp_path` provided by Pytest is cleaned up automatically after each test.

File: scripts/analysis/inference_flavor.py [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+    #     ],
+    #     avg_times_n_tasks_no_shared_model["Overall time"] / 60,
+    #     label="Overall time",
+    #     s=80,
+    #     marker="X",
+    #     facecolor="none",
+    #     edgecolor="red",
+    #     linewidths=1.3,
+    #     zorder=20,
+    # )
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/inference_flavor.py:241]: I removed them.

File: scripts/analysis/inference_flavor.py [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+    #     [str(n_tasks) for n_tasks in avg_times_n_tasks_shared_model["Number of nodes"]],
+    #     avg_times_n_tasks_shared_model["Overall time"] / 60,
+    #     label="Overall time",
+    #     s=80,
+    #     marker="X",
+    #     facecolor="none",
+    #     edgecolor="red",
+    #     linewidths=1.3,
+    #     zorder=20,
+    # )
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/inference_flavor.py:324]: Removed.

File: scripts/analysis/inference_flavor.py [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+                for column in target_columns:
+                    if column in df.columns:
+                        result = df.loc[df["comm_rank"] == "global", column].values[0]
+                    else:
+                        result = np.nan
+                    results[(data_set, number_of_tasks, model_seed)].append(result)
+
+    for dirpath, dirnames, filenames in os.walk(path_to_root):
+        for filename in filenames:
+            if filename.endswith(".out"):
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/inference_flavor.py:89]: Done.

File: scripts/analysis/inference_flavor.py [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+        "time_sec_forest_creation",
+        "time_sec_training",
+        "time_sec_all-gathering_model",
+        "time_sec_test",
+    ]
+
+    # Walk through the directory structure to find CSV files
+    for dirpath, dirnames, filenames in os.walk(path_to_root):
+        for filename in filenames:
+            if filename.endswith(".csv"):
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/inference_flavor.py:67]: Done.

File: scripts/analysis/weak_scaling_inference_flavor.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+    #     0.05, 0.95, energy_str, transform=ax6.transAxes, fontsize="x-small", verticalalignment="top", fontweight="bold"
+    # )
+    # Ensure the layout is tight so labels don't overlap
+    plt.tight_layout()
+    #
+    # Save the figure
+    plt.savefig(pathlib.Path(root_dir_no_shared_model) / f"{data_set}_weak_scaling.png")
+
+    # print(f"Overall energy consumed (no shared model): {(energy_no_shared_model/1000):.2f} kWh")
+    # print(f"Overall energy consumed (shared model): {(energy_shared_model/1000):.2f} kWh")
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/weak_scaling_inference_flavor.py]: Removed.

File: scripts/analysis/weak_scaling_inference_flavor.py [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+                    global_test_accuracy = df["accuracy_test"].values[0]
+
+                print(f"Global test accuracy: {global_test_accuracy}")
+                results[(data_set, number_of_tasks, model_seed)].append(
+                    global_test_accuracy
+                )
+
+    for dirpath, dirnames, filenames in os.walk(root_dir):
+        for filename in filenames:
+            if filename.endswith(".out"):
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/weak_scaling_inference_flavor.py]: Done.

File: scripts/analysis/weak_scaling_inference_flavor.py [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+    pd.DataFrame
+        The dataframe containing the results.
+    """
+    # Dictionary to store the values grouped by (dataset, number_of_tasks, dataseed)
+    results = defaultdict(list)
+
+    # Walk through the directory structure to find CSV files
+    for dirpath, dirnames, filenames in os.walk(root_dir):
+        for filename in filenames:
+            if filename.endswith(".csv"):
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/weak_scaling_inference_flavor.py]: Done.

File: scripts/analysis/weak_scaling.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+# Dictionary to store the values grouped by (dataset, number_of_tasks, dataseed)
+results = defaultdict(list)
+
+# The specific column you want to average in the CSV files
+target_columns = ["accuracy_global_test", "accuracy_test"]
+
+# Walk through the directory structure to find CSV files
+for dirpath, dirnames, filenames in os.walk(root_dir):
+    for filename in filenames:
+        if filename.endswith(".csv"):
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/weak_scaling.py:27]: Done.

File: scripts/analysis/weak_scaling.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+                global_test_accuracy = df["accuracy_test"].values[0]
+
+            print(f"Global test accuracy: {global_test_accuracy}")
+            results[(data_set, number_of_tasks, model_seed)].append(
+                global_test_accuracy
+            )
+
+for dirpath, dirnames, filenames in os.walk(root_dir):
+    for filename in filenames:
+        if filename.endswith(".out"):
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/weak_scaling.py:55]: Done.

File: scripts/analysis/weak_scaling.py
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/weak_scaling.py:1]: That is true and should be refactored in the future. As I only want some quick and dirty plotting of the results to check whether everything is ok, I would keep it for now for the sake of time and create an issue. I will work on this once all experiments ran through and I need to make more beautiful plots for the paper.

File: scripts/experiments/generate_strong_scaling_job_scripts.py
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/experiments/generate_strong_scaling_job_scripts.py:1]: Done.

File: specialcouscous/rf_parallel.py [hunk 13]
```diff
@@ -168,6 +169,24 @@ def _train_local_classifier(
...
+        Parameters
+        ----------
+        checkpoint_path : str | pathlib.Path
+            Path to the checkpoint directory containing the local model pickle files to load. There should only be one
+            valid checkpoint file for each rank, i.e., one file with the suffix "_rank_{comm.rank}.pickle}".
+        """
+        checkpoints = glob.glob(str(checkpoint_path) + "/*.pickle")
+        local_checkpoint = next(
+            ckpt for ckpt in checkpoints if f"_rank_{self.comm.rank}.pickle" in ckpt
+        )
```
[the remark this answers is not in the document]
Comment (mcw92) [reply specialcouscous/rf_parallel.py:189]: You are right, I also thought about the problem of multiple checkpoints in one directory. To deal with this, I enabled passing the considered checkpointed run's unique identifier in addition to the checkpoint path and additionally introduced exceptions when no or more than one checkpoint files are found. I also cleaned up the filtering.

File: scripts/analysis/weak_scaling_inference_flavor.py [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+
+pd.set_option("display.max_rows", None)
+
+# Get the root directory where results are stored from command line.
+root_dir = sys.argv[1]
+data_set = root_dir.split(os.sep)[-1]
+flavor = root_dir.split(os.sep)[-2].replace("_", " ")
+
+
+def get_results_df(
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/analysis/weak_scaling_inference_flavor.py:16]: See comment above: 

Will be refactored in the future. As I only want some quick and dirty plotting of the results to check whether everything is ok, I would keep it for now for the sake of time and create an issue. I will work on this once all experiments ran through and I need to make more beautiful plots for the paper.

File: scripts/experiments/generate_parallel_train_scripts.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+    log_n_features: int,
+    n_classes: int,
+    n_trees: int,
+    data_seed: int,
+    model_seed: int,
+    output_path: pathlib.Path,
+    submit: bool = False,
+) -> None:
+    """
+    Generate the job scripts for the inference flavor comparison experiments.
```
[the remark this answers is not in the document]
Comment (mcw92) [reply scripts/experiments/generate_parallel_train_scripts.py:17]: There is no real difference but these scripts more or less emerged due to the OOM issues when trying to run the inference comparison with our usual n6m4 and n7m3 baselines.
The inference comparison scripts generated with `generate_parallel_inference_comparison_job_scripts.py` consider smaller baselines as this series is only about comparing the inference flavors in terms of runtime and paralle overhead (and not so much about the accuracy). However, switching to smaller datasets here implied that I needed to rerun the normal weak scaling with more seeds for our usual baselines. The corresponding job scripts were generated with `generate_parallel_train_scripts.py`.
I cleaned up both scripts and clarified the docstrings.

File: scripts/examples/evaluate_from_checkpoint_parallel_synthetic.py [hunk 16]
```diff
@@ -27,13 +27,13 @@
 
     if comm.rank == 0:
         log.info(
-            "**********************************************************************\n"
-            "* Multi-Node Random Forest Classification of Balanced Synthetic Data *\n"
-            "**********************************************************************\n"
+            "********************** ************************************************\n"
+            "* Distributed Random Forest Classification of Balanced Synthetic Data *\n"
+            "*********************** ***********************************************\n"
```
Comment (fluegelk) [inline scripts/examples/evaluate_from_checkpoint_parallel_synthetic.py:32]: Are the spaces in the border intentional?

```suggestion
            "***********************************************************************\n"
            "* Distributed Random Forest Classification of Balanced Synthetic Data *\n"
            "***********************************************************************\n"
```

**Review by fluegelk — APPROVED** [review]
Looking good to me now 🎉 Only found one minor formatting thing in the evaluation example script

**Pull request merged**